src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/TypeNameBuilder.cs

```diff
@@ -0,0 +1,370 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+//  This TypeNameBuilder is ported from CoreCLR's original.
+//  It replaces the C++ bits of the implementation with a faithful C# port.
+
+using System.Collections.Generic;
+using System.Collections;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Diagnostics;
+
+namespace System.Reflection.Emit
+{
+    internal class TypeNameBuilder
+    {
+        private StringBuilder _str = new StringBuilder();
+        private int _instNesting;
+        private bool _firstInstArg;
+        private bool _nestedName;
+        private bool _hasAssemblySpec;
+        private bool _useAngleBracketsForGenerics;
+        private List<int> _stack = new List<int>();
+        private int _stackIdx;
+
+        public void OpenGenericArguments()
+        {
+            _instNesting++;
+            _firstInstArg = true;
+
+            if (_useAngleBracketsForGenerics)
+                Append('<');
+            else
+                Append('[');
+        }
+
+        public void CloseGenericArguments()
+        {
+            Debug.Assert(_instNesting != 0);
+
+            _instNesting--;
+
+            if (_firstInstArg)
+            {
+                _str.Remove(_str.Length - 1, 1);
+            }
+            else
+            {
+                if (_useAngleBracketsForGenerics)
+                    Append('>');
+                else
+                    Append(']');
+            }
+        }
+
+        public void OpenGenericArgument()
+        {
+            Debug.Assert(_instNesting != 0);
+
+            _nestedName = false;
+
+            if (!_firstInstArg)
+                Append(',');
+
+            _firstInstArg = false;
+
+            if (_useAngleBracketsForGenerics)
+                Append('<');
+            else
+                Append('[');
+
+            PushOpenGenericArgument();
+        }
+
+        public void CloseGenericArgument()
+        {
+            Debug.Assert(_instNesting != 0);
+
+            if (_hasAssemblySpec)
+            {
+                if (_useAngleBracketsForGenerics)
+                    Append('>');
+                else
+                    Append(']');
+            }
+
+            PopOpenGenericArgument();
+        }
+
+        public void AddName(string name)
+        {
+            Debug.Assert(name != null);
+
+            if (_nestedName)
+                Append('+');
+
+            _nestedName = true;
+
+            EscapeName(name!);
```

Comment:
```suggestion
            EscapeName(name);
```

`!` should not be necessary here.